From d1c27b34d1744336e56e7d160ed481aa48059894 Mon Sep 17 00:00:00 2001
From: Pierre-Hugues Husson <phh@phh.me>
Date: Fri, 24 Apr 2026 09:09:21 -0300
Subject: [PATCH] Don't abort when failing to get real-time priority

On some devices (like OP6), for unknown reason, trying to go to realtime
fails with EPERM.
There is no good reason to actually require real-time, so don't fail
when we don't get it
This fixes gabeldorsche on OP6
Not yet legacy bluetooth stack

Ported to android-16.0.0_r4 by bariccattion.
---
 system/btif/src/btif_a2dp_sink.cc       |  2 +-
 system/btif/src/btif_a2dp_source.cc     |  2 +-
 system/btif/src/btif_a2dp_source.cc.rej | 10 ++++++++++
 system/osi/src/alarm.cc                 | 12 +++++++++++-
 system/stack/btu/main_thread.cc         |  2 +-
 5 files changed, 24 insertions(+), 4 deletions(-)
 create mode 100644 system/btif/src/btif_a2dp_source.cc.rej

diff --git a/system/btif/src/btif_a2dp_sink.cc b/system/btif/src/btif_a2dp_sink.cc
index d20be9dc12..c07249e06a 100644
--- a/system/btif/src/btif_a2dp_sink.cc
+++ b/system/btif/src/btif_a2dp_sink.cc
@@ -203,7 +203,7 @@ bool btif_a2dp_sink_init() {
   /* Schedule the rest of the operations */
   if (!btif_a2dp_sink_cb.worker_thread.EnableRealTimeScheduling()) {
 #if defined(__ANDROID__)
-    log::fatal("Failed to increase A2DP decoder thread priority");
+    log::error("Failed to increase A2DP decoder thread priority");
 #endif
   }
   btif_a2dp_sink_cb.worker_thread.DoInThread(base::BindOnce(btif_a2dp_sink_init_delayed));
diff --git a/system/btif/src/btif_a2dp_source.cc b/system/btif/src/btif_a2dp_source.cc
index ae00fa4323..88a6ba0fe8 100644
--- a/system/btif/src/btif_a2dp_source.cc
+++ b/system/btif/src/btif_a2dp_source.cc
@@ -335,7 +335,7 @@ bool btif_a2dp_source_init(void) {
   if (com_android_bluetooth_flags_a2dp_source_null_fixed_queue()) {
     if (!btif_a2dp_source_thread.EnableRealTimeScheduling()) {
 #if defined(__ANDROID__)
-      log::fatal("unable to enable real time scheduling");
+log::error("unable to enable real time scheduling");
 #endif
     }
   }
diff --git a/system/btif/src/btif_a2dp_source.cc.rej b/system/btif/src/btif_a2dp_source.cc.rej
new file mode 100644
index 0000000000..ba76a1ff4b
--- /dev/null
+++ b/system/btif/src/btif_a2dp_source.cc.rej
@@ -0,0 +1,10 @@
+diff a/system/btif/src/btif_a2dp_source.cc b/system/btif/src/btif_a2dp_source.cc	(rejected hunks)
+@@ -438,7 +438,7 @@ static void btif_a2dp_source_startup_delayed() {
+   log::info("state={}", btif_a2dp_source_cb.StateStr());
+   if (!btif_a2dp_source_thread.EnableRealTimeScheduling()) {
+ #if defined(__ANDROID__)
+-    log::fatal("unable to enable real time scheduling");
++    log::error("unable to enable real time scheduling");
+ #endif
+   }
+   if (!bluetooth::audio::a2dp::init(get_main_thread(), &a2dp_stream_callbacks,
diff --git a/system/osi/src/alarm.cc b/system/osi/src/alarm.cc
index ee62645221..4b1838517e 100644
--- a/system/osi/src/alarm.cc
+++ b/system/osi/src/alarm.cc
@@ -687,7 +687,17 @@ static bool timer_create_internal(const clockid_t clock_id, timer_t* timer) {
   sigevent.sigev_notify = SIGEV_THREAD;
   sigevent.sigev_notify_function = timer_callback;
   sigevent.sigev_notify_attributes = &thread_attr;
-  if (timer_create(clock_id, &sigevent, timer) == -1) {
+
+  int ret = timer_create(clock_id, &sigevent, timer);
+  if (ret == -1) {
+       log::error("{} failed to create timer with RT err {}... Try again without RT", __func__, strerror(errno));
+      // Recreate timer without RT priority
+      memset(&sigevent, 0, sizeof(sigevent));
+      sigevent.sigev_notify = SIGEV_THREAD;
+      sigevent.sigev_notify_function = (void (*)(union sigval))timer_callback;
+      ret = timer_create(clock_id, &sigevent, timer);
+  }
+  if (ret == -1) {
     log::error("unable to create timer with clock {}: {}", clock_id, strerror(errno));
     if (clock_id == CLOCK_BOOTTIME_ALARM) {
       log::error(
diff --git a/system/stack/btu/main_thread.cc b/system/stack/btu/main_thread.cc
index b24041a019..ac5a320bf0 100644
--- a/system/stack/btu/main_thread.cc
+++ b/system/stack/btu/main_thread.cc
@@ -69,7 +69,7 @@ void main_thread_start_up() {
   }
   if (!main_thread.EnableRealTimeScheduling()) {
 #if defined(__ANDROID__)
-    log::fatal("unable to enable real time scheduling");
+    log::error("unable to enable real time scheduling");
 #else
     log::error("unable to enable real time scheduling");
 #endif
-- 
2.53.0.windows.3

